lib/types.nix

```
@@ -373,6 +373,14 @@ rec {
       merge = mergeEqualOption;
     };
 
+    hexStr = mkOptionType {
+      name = "hexStr";
+      description = "hexadecimal value represented as string";
+      descriptionClass = "noun";
+      check = x: str.check x && builtins.match "[0-9a-f]+" x != null;
```

In line with https://github.com/NixOS/nixpkgs/pull/301039#discussion_r1575343406, something like this:

```suggestion
      check = x:
        str.check x
        && builtins.match "[0-9a-f]+" x != null
        # Usually hex digits are used to represent bytes.
        # A single hex digit only represents half a byte,
        # so if we have an odd number, it's likely a mistake.
        && mod (stringLength x) 2 == 0;
```

Writing this, I'm actually not sure if it's worth doing this, the error message when you get the type wrong is really not very helpful in such a case.

---

Right, this only outputs a bool.
What if we check it in the merge function instead, like the other, good errors?
Technically that breaks `either hexStr x` for odd-length definitions, but that is probably ok in practice.

Perhaps the module system could invoke the merge function instead of producing a generic error message? Or falling back to the generic message if the merge evaluates (`if ! t.check v then seq merged (throw genericMsg) else merged`).
(Just an idea, not a request)

---

I think the proper solution is to extend `.check` to allow custom error messages.

To move forward with this PR though, I don't really care about having this check or not. I'd like to leave this decision up to @Janik-Haag.